[4-6단계 - 방탈출 사용자 예약] 젠슨(이성우) 미션 제출합니다.

CodingMasterLSW opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
단일 책임, 메서드 15라인 등 어느정도 잘 지켰다고 생각합니다.


## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

인프라 담당이었어서, 스프링에 대한 이해도는 매우 낮습니다.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

안녕하세요 베루스, 미션 마감 기한이 얼마 남지 않아 테스트 코드를 거의 작성하지 못 했습니다.
시간 부족으로 인해 우선 PR 제출하고, 이후에 테스트 코드 차근차근 작성하겠습니다.

Q. 기능이 늘어날 수록 클래스 관리가 헷갈려 패키지 구조를 잘게 나눠봤습니다. 이에 피드백을 받고 싶습니다!

PR 보낸 후 추가로 할 작업
1) 예외처리 세분화 + 예외에 따른 상태코드 던지기
2) 테스트 코드 작성


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

CodingMasterLSW committed:
feat: Member 테이블 생성 및 테이블 연관관계 수정

CodingMasterLSW committed:
feat: member 저장 API 생성 및 기존 코드 일부 주석 처리

CodingMasterLSW committed:
feat: 로그인 페이지 응답 구현

CodingMasterLSW committed:
feat: 로그인 요청에 응답하는 API 생성

- request 값을 통해 유저 조회
- 조회한 멤버로 토큰 생성
- 토큰을 쿠키에 담아 반환

CodingMasterLSW committed:
refactor: 회원가입 endpoint 변경 및 4단계 코드 주석 해제

CodingMasterLSW committed:
feat: 인증 정보 조회 API 생성

CodingMasterLSW committed:
feat: 로그아웃 API 추가

CodingMasterLSW committed:
feat: 로그아웃이 성공할 경우, 로그인 페이지로 리다이렉션 하도록 변경

CodingMasterLSW committed:
refactor: Login과 Member 분리

- LoginController/MemberController
- AuthService/MemberService 로 분리하였음

CodingMasterLSW committed:
refactor: responsDto, requestDto 패키지 정리

request, response를 따로 분리하였음

CodingMasterLSW committed:
refactor: 유저 객체 도입으로 인한 예약 DAO 수정

CodingMasterLSW committed:
test: Member 객체 도입으로 인한 기존 repository 테스트 코드 수정

CodingMasterLSW committed:
refactor: 예약 생성 기능 변경 및 기존 Reservation 코드 리팩토링

- ArgumnetResolver 도입을 통해 쿠키를 이용해 예약 생성을 하도록 변경하였음
- Member 도입으로 인해 돌아가지 않는 테스트 코드들을 수정하였음

CodingMasterLSW committed:
refactor: 관리자가 예약 생성 시, 로그인한 사용자 정보를 활용하도록 리팩터링을 진행

- 모든 유저 조회 API 추가
- 유저 ID를 통한 유저 조회 API 추가
- 쿠키를 이용해 예약 하도록 관리자 예약 페이지 수정

CodingMasterLSW committed:
refactor: 예약 서비스 기능단위별 메서드 분리 및 DTO 네이밍 변경

CodingMasterLSW committed:
refactor: 예약이 존재하는지 검증하는 퀴리문 수정으로 인한 리팩토링

- count값 반환 -> boolean 값 반환으로 수정

CodingMasterLSW committed:
refactor: Member의 역할 추가

- Role을 도입해 Admin, Member를 구분지었음

CodingMasterLSW committed:
feat: admin 페이지에 접근 권한 사장님으로 제한

CodingMasterLSW committed:
feat: 예약 목록 검색 API 추가

CodingMasterLSW committed:
refactor: 패키지 정리

- global, domain으로 분리 하였음

CodingMasterLSW committed:
refactor: repositoty 계층 삭제

CodingMasterLSW committed:
refactor: delete 로직 수정

- 존재하면 삭제 -> 삭제 여부 확인으로 로직을 수정하였음

CodingMasterLSW committed:
refactor: 이미 존재하는 회원이면, 회원가입 할 수 없도록 수정

CodingMasterLSW committed:
refactor: 예약 존재여부를 포함한 예약 시간 반환 로직 수정

- DAO에서의 쿼리문을 통해 boolean 값 반환 -> Service에서 boolean 값 추가하도록 로직 수정

CodingMasterLSW committed:
refactor: jwt 생성 및 검증 로직 수정

- claim에 name, role 추가
- jwt 버전 변경으로 인한 레거시 코드 수정

CodingMasterLSW committed:
refactor: LoginMemberArgumentResolver를 활용한 리팩토링

- Controller에서 ArgumentResolver를 활용하도록 수정하였음

CodingMasterLSW committed:
test: 관리자 페이지 테스트코드 임시 주석처리

CodingMasterLSW committed:
remove: 레거시 코드 삭제 및 주석 추가

CodingMasterLSW committed:
fix: 중복된 예약 조회에서의 오류 수정

- 테마 조회를 추가하였음

CodingMasterLSW commented on src/test/java/roomescape/service/ReservationServiceTest.java:
```diff
@@ -1,87 +1,144 @@
+            .doesNotThrowAnyException();
+    }
+
+    @Test
+    @DisplayName("존재하지 않는 예약ID를 삭제하려고 할 경우, 예외가 발생해야 한다.")
+    void not_exist_theme_id_delete_then_throw_exception() {
+        Reservation reservation = createReservation(createMember(), createTheme(),
+            createReservationTime());
+        long savedId = reservationDao.save(reservation);
+        reservation.setId(savedId);
+        assertThatCode(() -> reservationService.deleteReservation(savedId))
+            .doesNotThrowAnyException();
     }
```
좋은 지적이야. 수정했어

CodingMasterLSW commented on src/main/java/roomescape/domain/reservationtime/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,64 @@
+    public void findById(Long id) {
+        reservationTimeDao.findById(id)
+            .orElseThrow(() -> new IllegalArgumentException("해당 ID의 예약시간을 찾을 수 없습니다"));
+    }
```
테스트에서만 사용하고 있어서 삭제하였음

CodingMasterLSW commented on src/main/java/roomescape/domain/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,63 @@
+        @RequestBody AdminReservationRequestDto adminReservationRequestDto, Member member) {
+        return reservationService.saveReservationOfAdmin(adminReservationRequestDto,
+            member.getId());
+    }
```
이건 인터셉터에서 하고있단다.

CodingMasterLSW commented on src/main/java/roomescape/domain/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,63 @@
+            member.getId());
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteReservation(@PathVariable(name = "id") Long id) {
+        reservationService.deleteReservation(id);
+    }
```
이건 생각해볼 부분이네. 
reservation, reservationTime의 추가, 삭제를 관리자만 접근할 수 있도록 고려해봐야겠어.

CodingMasterLSW commented on src/main/java/roomescape/domain/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,130 @@
+        validateDateTimeAndSaveReservation(reservation, themeId);
+        return reservation;
+    }
+
```
이걸 빨리 봤어야했는데... 이제 너 말을 잘 읽어볼게 토끼야

CodingMasterLSW commented on src/main/java/roomescape/domain/reservation/dao/JdbcReservationDaoImpl.java:
```diff
@@ -0,0 +1,211 @@
+
+    @Override
+    public List<Reservation> findAll() {
+        String sql = """
+            select
+                        m.name as member_name,
+                        m.id as member_id,
+                        m.role as member_role,
+                        m.email as member_email,
+                        m.password as member_password,
+                        r.id as reservation_id,
+                        r.date as reservation_date,
+                        rt.id as time_id,
+                        rt.start_at,
+                        t.id as theme_id,
+                        t.name as theme_name,
+                        t.description as theme_description,
+                        t.thumbnail as theme_thumbnail
+            from reservation r
+            inner join member m on r.member_id = m.id
+            inner join reservation_time rt on r.time_id = rt.id
+            inner join theme t on r.theme_id = t.id  
+            """;
+
+        return jdbcTemplate.query(sql,
+            (resultSet, RowNum) ->
+                new Reservation(resultSet.getLong("reservation_id"),
+                    createMember(resultSet),
+                    createReservationDate(resultSet),
+                    createReservationTime(resultSet),
+                    createTheme(resultSet)
+                ));
+    }
+
+    private ReservationDate createReservationDate(ResultSet resultSet) throws SQLException {
+        return new ReservationDate(LocalDate.parse(resultSet.getString("date")));
+    }
```
start_at은 별 문제가 없어보이는데?

CodingMasterLSW commented on src/main/java/roomescape/domain/reservation/dao/JdbcReservationDaoImpl.java:
```diff
@@ -0,0 +1,211 @@
+            where r.id = ?
+            """;
+        try {
+            Reservation reservation = jdbcTemplate.queryForObject(query,
+                (resultSet, rowNum) ->
+                    new Reservation(
+                        resultSet.getLong("id"),
+                        createMember(resultSet),
+                        createReservationDate(resultSet),
+                        createReservationTime(resultSet),
+                        createTheme(resultSet)
+                    ),
+                id
```
수정했어.

CodingMasterLSW commented on src/test/java/roomescape/dao/FakeReservationDaoImpl.java:
```diff
@@ -38,24 +40,26 @@ public Optional<Reservation> findById(Long id) {
-            .filter(reservation -> reservation.getReservationDate().equals(date))
-            .filter(reservation -> reservation.getTimeId() == timeId)
-            .count();
+    public List<Reservation> findOf(String dateFrom, String dateTo, Long memberId, Long themeId) {
+        return List.of();
```
추가할게.

CodingMasterLSW commented on src/test/java/roomescape/dao/FakeReservationDaoImpl.java:
```diff
@@ -38,24 +40,26 @@ public Optional<Reservation> findById(Long id) {
-        return (int) reservations.stream()
-            .filter(reservation -> reservation.getId().equals(id))
-            .count();
+    public boolean existReservationByTime(Long id) {
+        return reservations.stream()
+            .anyMatch(reservation -> reservation.getId().equals(id));
     }
```
그러네... 수정할게

CodingMasterLSW committed:
refactor: 레거시 코드 삭제 및 논리적 오류가 있는 테스트 코드 수정

CodingMasterLSW committed:
remove: 붎필요한 print문 삭제

CodingMasterLSW committed:
refactor: 예외처리 세분화

verus-j commented on src/test/java/roomescape/service/ReservationServiceTest.java:
```diff
@@ -1,87 +1,144 @@
 
     @DisplayName("이미 존재하는 날짜와 시간에 예약하려고 하면, 예외가 발생해야 한다.")
     @Test
     void already_exist_date_time_save_reservation_then_throw_exception() {
```
테스트가 실패하네요~ 에러문구 확인부탁드려요!

CodingMasterLSW replied:
수정했습니다!

verus-j commented on LICENSE:
```diff
@@ -2,7 +2,7 @@ MIT License
 
```
어떤 부분에서 클래스 관리가 어려웠을까요? Controller나 Service는 패키지로 나누지 않아도 Prefix로 인해 충분히 구분될 것 같은데 어떤 부분에서 클래스 관리가 어려웠는지 궁금하네요~

CodingMasterLSW replied:
우선 패키지를 나누게 된 계기는 DTO class의 증가였습니다.
Request/Response로 나눴는데도 많은 클래스들이 존재했고 추후에 기능이 추가된다면, 분리해야 하는 상황이 언젠가는 온다고 판단했습니다.

어떻게 하면 패키지를 잘 구분할 수 있을까? 고민하다가 global/ domain 으로 먼저 구분을 해봤습니다.
이후에 각 도메인 별로 controller, service, repository... 와 같은 원래 패키지 형태를 사용해봤습니다.
현 시점에서는 controller, service, repository 패키지 전부 삭제하고, class 단위로 관리해도 좋아보입니다.

ex)
- reservation
      - reservation
      - reservationController
      - reservationService
      - reservationDao
      - dto
              - ...response
               - ...request

습관적으로 controller, service, repository 패키지를 나누어 관리했었는데, 현 시점에서는 불필요하다고 생각하시는걸까요?
딱히 controller/service/repository 를 구분한 의도가 있지는 않습니다.

verus-j replied:
reservation.controller, reservationtime.controller 와 같은 도메인 하위의 controller, service, dao 패키지에는 하나의 클래스(dao는 인터페이스와 구현체)만 있어서 오히려 너무 세분화한것 아닌가라는 의문이 들었습니다. dto의 증가라면 아래와 같이 dto만 패키지를 나눠서 관리할 수 있지 않았을까요??

* dto
  * reservation
    * ...
  * reservationtime
    * ...
  * ...

verus-j commented on src/main/java/roomescape/domain/member/dao/JdbcMemberDaoImpl.java:
```diff
@@ -0,0 +1,120 @@
+
+    @Override
+    public Optional<Member> findByEmailAndPassword(String email, String password) {
+        String query = """
+            select *
+             from member
+             where email = ? AND password = ?;
+            """;
+        try {
+            return Optional.ofNullable(jdbcTemplate.queryForObject(query,
+                (resultSet, rowNum) -> {
+                    return Member.createMember(
+                        resultSet.getLong("id"),
+                        Role.convertFrom(resultSet.getString("role")),
+                        resultSet.getString("name"),
+                        resultSet.getString("email"),
+                        resultSet.getString("password")
+                    );
+                }, email, password));
+        } catch (EmptyResultDataAccessException e) {
+            return Optional.empty();
+        }
```
평문 비밀번호로 검증해도 괜찮을까요? 한번 고민해보면 좋을것 같습니다!

CodingMasterLSW replied:
평문 비밀번호로 검증은 좋지 않은 방법이라고 생각합니다.
객체로 생성되거나, DB에서 저장되는 시점에 secretKey + 암호화 알고리즘 + 비밀번호 를 조합한 이후에 DB에서 관리하는게 보안적인 부분에서 안정적이라고 생각했습니다.

이에 대해 궁금한 부분은, Client에서 전송되는 시점에서는 따로 암호화가 필요하지 않나? 입니다.
만약 Client에서  전송되는 시점에 암호화되어 보내진다면, 암호화된 값만 저장하면 되지 않을까? 라고 생각했습니다.

보통, Client -> 암호화 -> Server의 흐름이 이상적인가요?
아니면 Client -> Server (암호화 ) 의 흐름이 이상적인가요?

혹은 두 곳 전부 암호화를 따로 진행하는 것이 이상적인가요?? 궁금합니다!

verus-j replied:
client에서도 암호화를 해서 보내지만 클라이언트에서 암호화된 값을 그대로 사용한다면 악의적인 사용자가 직접 rest call을 호출해 로그인을 할 수 있는 위험이 있습니다. 보통 서버쪽에서 암호화하여 DB에 저장하고 클라이언트 쪽은 데이터 전송을 위해 암호화만 진행합니다.

verus-j commented on src/main/java/roomescape/domain/member/dto/response/MemberResponseDto.java:
```diff
@@ -0,0 +1,11 @@
+
+public record MemberResponseDto(String token) {
+
+    // TODO : 추후에 token을 반환하도록 리팩토링 하기
```
여기는 어떤 의미의 주석일까요? 그리고 필드명이 token이 아니라 name이 맞는걸까요?

CodingMasterLSW replied:
잘못된 주석입니다. 삭제 진행하겠습니다.

verus-j commented on src/main/java/roomescape/domain/member/model/Member.java:
```diff
@@ -0,0 +1,74 @@
+        return new Member(Role.USER, name, email, password);
+    }
+
+    /**
+     * 사장님 로그인 페이지가 추가된다면, 사용할 정적 팩토리 메서드 코드
+     */
+    public static Member createAdminWithId(Long id, String name, String email, String password) {
+        return new Member(id, Role.ADMIN, name, email, password);
+    }
```
사장님 로그인 페이지가 언제 추가되는걸까요?

CodingMasterLSW replied:
오타입니다. (로그인 페이지 -> 회원가입 페이지)

미션에서는 사장님 로그인 페이지를 요구하지 않았지만, 
비즈니스 로직의 흐름상 사장님 회원가입 페이지는 따로 존재해야된다고 생각했습니다.

verus-j replied:
아직 불필요한 기능이라 생각되어 언급했습니다. 다음 단계에서 다른 크루가 젠슨의 코드로 미션을 진행할 수 있는데, 사장님 로그인에 대해 물어보지 않을까요?

verus-j commented on src/main/java/roomescape/domain/member/model/Member.java:
```diff
@@ -0,0 +1,74 @@
+        return new Member(id, Role.ADMIN, name, email, password);
+    }
+
+    public static Member createAdmin(String name, String email, String password) {
```
여기는 사용하지 않는 메서드네요

CodingMasterLSW replied:
위에서 언급한 것과 마찬가지로, 추후 확장성을 위해 남겨놨습니다.
비즈니스 로직 흐름상, 사장님 회원가입 페이지는 따로 존재햐아 된다고 생각했고, 미래를 위해 남겨놓은 코드였습니다.
(사장님 로그인 페이지가 없다면, 직접 SQL문을 사용해 데이터를 저장해야 하는 상황이기에, 필요하다고 생각했습니다.)

verus-j commented on src/main/java/roomescape/domain/reservation/dao/JdbcReservationDaoImpl.java:
```diff
@@ -0,0 +1,211 @@
+
+    @Override
+    public List<Reservation> findAll() {
+        String sql = """
+            select
+                        m.name as member_name,
+                        m.id as member_id,
+                        m.role as member_role,
+                        m.email as member_email,
+                        m.password as member_password,
+                        r.id as reservation_id,
+                        r.date as reservation_date,
+                        rt.id as time_id,
+                        rt.start_at,
+                        t.id as theme_id,
+                        t.name as theme_name,
+                        t.description as theme_description,
+                        t.thumbnail as theme_thumbnail
+            from reservation r
+            inner join member m on r.member_id = m.id
+            inner join reservation_time rt on r.time_id = rt.id
+            inner join theme t on r.theme_id = t.id  
+            """;
```
굉장히 복잡한 쿼리네요. 이전에 피드백을 드렸는지 기억이 가물가물한데, 심플한 쿼리를 여러개 구성해 서비스에서 해당 로직을 구현해보는건 어떻게 생각하나요?

verus-j reviewed (changes requested):
안녕하세요 젠슨~
4-6단계 진행하느라 고생많으셨습니다.
피드백 몇개 남겨놓았고, 미리 얘기해주신데로 테스트 코드 작성마저 진행해주시면 좋을 것 같습니다!

CodingMasterLSW committed:
refactor: 레거시 TODO 삭제 및 테스트 코드 수정

CodingMasterLSW committed:
test: memberService 테스트 코드 작성

CodingMasterLSW committed:
test: MemberDao 테스트코드 작성

CodingMasterLSW committed:
test: Login 테스트 코드 작성

CodingMasterLSW committed:
test: Jwt 생성 및 검증 테스트 코드 작성

CodingMasterLSW committed:
refactor: Login 도메인 패키지 생성 및 메서드 네이밍 수정

- CheckResponseDto -> LoginResponseDto
- TokenRequest -> LoginRequest
- AuthService -> LoginService

CodingMasterLSW committed:
feat: 관리자 회원가입 기능 추가

CodingMasterLSW commented:
안녕하세요 베루스, 
이번 리뷰 요청에서는 추가로 테스트 코드 추가 및 관리자 회원가입 페이지를 추가했습니다.

커멘트에 답글 달아놨습니다 감사합니다!

verus-j reviewed (approved):
안녕하세요 젠슨
미션 진행하시느라 고생많으셨습니다.
다음 미션 진행을 위해 여기서 이만 머지하겠습니다
남은 미션도 화이팅입니다~
